server/WFCORE-6750-unstable-api-annotation-scanner.adoc

```diff
@@ -0,0 +1,243 @@
+
+== Backwards Compatibility
+
+// Does this enhancement affect backwards compatibility with previously released
+// versions of WildFly?
+// Can the identified incompatibility be avoided?
+
+=== Default Configuration
+
+Since this is a preview stability feature, this will not be added to the default configuration.
+
+=== Importing Existing Configuration
+
+Importing existing configuration will be fine, as this is a new feature.
+
+=== Deployments
+
+If the feature is enabled, deployments will be scanned for usage of API elements existing in the mentioned indices.
+
+=== Interoperability
+
+N/A
+
+== Security Considerations
+
+////
+Identification if any security implications that may need to be considered with this feature
+or a confirmation that there are no security implications to consider.
+////
+I don't believe there are any security issues.
+
+== Test Plan
+
+Tests will be added in both WildFLy and WildFly Core
+
+WildFly Core:
+
+* The WildFly core feature pack is only used internally for testing, so we will not scan for annotations indicating unstable API with the plugin here. Also, it should be noted that the  usage of such annotations in third party components is outside of our control, and likely to change (e.g. an annotated member might 'mature' to no longer have the annotation, or the member might be removed in a future release). Meanwhile, the tests for the scanning mechanism should be in WildFly Core since that is where the deployment unit processors live.
+** Thus, a test feature pack will be created offering an API (including its own annotations for indicating unstable API) and a subsystem providing this API on the classpath.
+** The feature pack will be indexed and the indices will be added to the `{indexModule}` module.
+** The test will check that
+*** the index file is added to the `${indexModule}` module, and to the list in the module's `index.txt`, as outlined in the requirements
+*** when deploying an application using annotated API elements from the test feature pack API, usage of those is reported in line with the configuration in the `subsystem=core-management/service=unstable-api-annotations` resource (as outlined in the requirements). We will test both with
+**** `level=log` - and inspect that the log messages pick up all the usages of annotated API elements.
+**** `level=error` - and inspect that the application fails to deploy, and that the error message picks up all the usages of annotated API elements.
+** The test will run at the `preview` stability level
+
+WildFly:
+
+* Since we have done the main testing of the mechanism in WildFly Core, we will not test that here
+* A test will be added to `testsuite/integration/basic` which runs in an execution provisioning both `wildfly-ee-feature-pack` and `wildfly-galleon-pack`. It will test
```

Comment:
This needs to be in testsuite/integration/microprofile (to be renamed 'expansion').

Use of wildfly-galleon-pack is not allowed in testsuite/integration/basic.

Reply:
Also s/wildfly-ee-feature-pack/wildfly-ee-galleon-pack

or, my minor preference

"both the 'wildfly-ee' and 'wildfly' feature packs."

Reply:
Good point, it will simplify the test since in basic there were two cases
- the full zip which tests everything
- the galleon provisioned server which don't have the wildfly FP